File: koka/lib/std/time/timer-inline.c
/*---------------------------------------------------------------------------
  Copyright 2020-2021, Microsoft Research, Daan Leijen.

  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the LICENSE file at the root of this distribution.
---------------------------------------------------------------------------*/

static kk_std_core_types__tuple2 kk_timer_ticks_tuple(kk_context_t* ctx) {
  kk_duration_t d = kk_timer_ticks(ctx);
  // the conversion has about 15 digits of precision
  // we cannot do this more precisely as the api expects the fraction between 0.0 and 2.0 (for leap seconds).
  double secs = (double)d.seconds;
  double frac = (double)d.attoseconds * 1e-18;
  return kk_std_core_types__new_Tuple2( kk_double_box(secs,ctx), kk_double_box(frac,ctx), ctx );
}

static double kk_timer_dresolution(kk_context_t* ctx) {
  int64_t asecs = kk_timer_resolution(ctx);
  return (double)asecs * 1e-18;
}

File: koka/lib/std/time/timer-inline.cs
/*---------------------------------------------------------------------------
  Copyright 2016-2021, Microsoft Research, Daan Leijen.

  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the LICENSE file at the root of this distribution.
---------------------------------------------------------------------------*/
using System.Diagnostics;

static class _Timer {
  private static Stopwatch ticker = Stopwatch.StartNew();
  private static long frequency = (Stopwatch.Frequency > 0 ? Stopwatch.Frequency : 1000);
  private static double resolution = 1.0 / (double)frequency;

  public static __std_core_types._Tuple2_<double,double> Ticks() {
    long ticks = ticker.ElapsedTicks;
    // maintain precision ..
    double secs  = (double)(ticks / frequency);
    double fsecs = (double)(ticks % frequency) * resolution;
    return new __std_core_types._Tuple2_<double,double>(secs,fsecs);
  }

  public static double TicksResolution() {
    return resolution;
  }
}
